### PR TITLE
Feature/multiple primitive mutations

### PR DESCRIPTION
**Pull request by nklapste**
Adds the ability for operator mutations (and possibly other primitive mutations) to have multiple alternatives.

This should be able to address the request of additional mutations for assignment operators as noted in #124.

For example the python statement:

```python
x += 2
```
could now be mutated to both the originally supported mutmut mutation:
```python
x -= 2
```
and a new alternative mutation:
```python
x = 2
```

### CODE REVIEW

**Comment by boxed** [comment]
We should maybe check how this approach compares to what they did in https://github.com/kc7zep/uw-mutmut

**Comment by nklapste** [comment]
> 
> 
> We should maybe check how this approach compares to what they did in https://github.com/kc7zep/uw-mutmut

They use a dictionary system, I'm using a list (maybe a set is more appropriate, though the order consistency of lists may make debugging easier).

Both unload the new type in a similar way (applying the mutations in a for loop extractor), it seems that we mostly differ in datatype implementation.

**Comment by boxed** [comment]
Cool! I agree that a list is more appropriate.

**Comment by nklapste** [comment]
@boxed unless you have some suggestions on some good alternate mutations to add (possibly with other primitive operations such as `+` and `-`), I am ready for review and merging this into master.

**Comment by nklapste** [comment]
@boxed nevermind that last comment I found a issue if you modify the integration tests within `test_main.py` to include/modify the lines:

```python
file_to_mutate_lines = [
    "def foo(a, b):",
    "    return a < b",
    "c = 1",
    "c += 1",
    "e = 1",
    "f = 3",
    "d = dict(e=f)",
]
```
I get a `AssertionError` noting that no mutations were preformed (this however is not the case).

**Comment by nklapste** [comment]
> 
> 
> @boxed nevermind that last comment I found a issue if you modify the integration tests within `test_main.py` to include/modify the lines:
> 
> ```python
> file_to_mutate_lines = [
>     "def foo(a, b):",
>     "    return a < b",
>     "c = 1",
>     "c += 1",
>     "e = 1",
>     "f = 3",
>     "d = dict(e=f)",
> ]
> ```
> 
> I get a `AssertionError` noting that no mutations were preformed (this however is not the case as the later `g: int = 2` to `g: int = None` mutation is still applied).
This should now be resolved. 

Turns out a mutation stopping shortcut conditional was not included in my iteration of alternate mutations. 

Additionally, this fix-up also includes some simplifications within `__init__.py` and on the `Context` class within `__init__.py` as the field `number_of_performed_mutations` was removed (it is redundant as it is essentially a copy of length of `context.performed_mutation_ids`)

**Comment by nklapste** [comment]
I resolved the issues present with no mutations being preformed. Turned out the alternate mutations were creating a unobtainable state between mutation generation (where all mutations are applied in order) and mutation execution (one mutation is applied, others are skipped).

The last applied mutation of a node in mutation generation **can** influence mutations on next nodes. In mutation execution (i.e. only applying one mutant) care must be taken to ensure that it can still be obtained despite mutations present in mutation generation now being missing.

After reversing the application of alternate mutations (thus, making the last applied mutation the "old-school" mutmut mutation) behavior seems to be stable. 

@boxed feel free to review. I haven't been able to raise any major issues with my additions so far.

File: mutmut/__init__.py [hunk 1]
```diff
@@ -537,17 +540,26 @@ def mutate_node(node, context):
...
-
-            # this is just an optimization to stop early
-            if context.number_of_performed_mutations and context.mutation_id != ALL:
-                return
+            if isinstance(new, list) and not isinstance(old, list):
+                # multiple mutations
+                news = new
+            else:
+                # one mutation
+                news = [new]
```
Comment (boxed) [inline mutmut/__init__.py]: "news" as plural of "new" is a bit cute, but not correct English! new_list would be better.

File: mutmut/__init__.py [hunk 2]
```diff
@@ -572,19 +582,6 @@ def mutate_list_of_nodes(node, context):
 
         mutate_node(child_node, context=context)
 
-        # this is just an optimization to stop early
-        if context.number_of_performed_mutations and context.mutation_id != ALL:
-            return
```
Comment (boxed) [inline mutmut/__init__.py]: Are you sure you want to get rid of this early return?

**Comment by boxed** [comment]
Some minor nitpicks, but overall looks good!

**Comment by nklapste** [comment]
Currently, I believe a issue with tri-declaritivie is causing a issue with the python2.y and pypy builds:

See the following stack trace:
```
F
tests\test_main.py:198 (test_simple_apply)
filesystem = local('c:\\users\\nklap\\appdata\\local\\temp\\pytest-of-nklap\\pytest-5\\test_fs4')

    def test_simple_apply(filesystem):
>       result = CliRunner().invoke(climain, ['run', '--paths-to-mutate=foo.py', "--test-time-base=15.0"], catch_exceptions=False)

C:\Users\nklap\PycharmProjects\mutmut\tests\test_main.py:200: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\nklap\PycharmProjects\mutmut\venv27\lib\site-packages\click\testing.py:326: in invoke
    cli.main(args=args or (), prog_name=prog_name, **extra)
C:\Users\nklap\PycharmProjects\mutmut\venv27\lib\site-packages\click\core.py:717: in main
    rv = self.invoke(ctx)
C:\Users\nklap\PycharmProjects\mutmut\venv27\lib\site-packages\click\core.py:956: in invoke
    return ctx.invoke(self.callback, **ctx.params)
C:\Users\nklap\PycharmProjects\mutmut\venv27\lib\site-packages\click\core.py:555: in invoke
    return callback(*args, **kwargs)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__main__.py:75: in wrapper
    f(*args, **kwargs)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__main__.py:254: in climain
    post_mutation, use_patch_file, paths_to_exclude))
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__main__.py:402: in main
    add_mutations_by_file(mutations_by_file, filename, _exclude, dict_synonyms)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__main__.py:722: in add_mutations_by_file
    mutations_by_file[filename] = list_mutations(context)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__init__.py:595: in list_mutations
    mutate(context)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__init__.py:482: in mutate
    mutate_list_of_nodes(result, context=context)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__init__.py:583: in mutate_list_of_nodes
    mutate_node(child_node, context=context)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__init__.py:520: in mutate_node
    mutate_list_of_nodes(node, context=context)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__init__.py:583: in mutate_list_of_nodes
    mutate_node(child_node, context=context)
C:\Users\nklap\PycharmProjects\mutmut\mutmut\__init__.py:541: in mutate_node
    children=getattr(node, 'children', None),
C:\Users\nklap\PycharmProjects\mutmut\venv27\lib\site-packages\tri_declarative\__init__.py:357: in evaluate
    callee_parameters = get_signature(func_or_value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

func = <function keyword_mutation at 0x00000000050C2128>

    def get_signature(func):
        """
            :type func: Callable
            :rtype: str
        """
        try:
            return object.__getattribute__(func, '__tri_declarative_signature')
        except AttributeError:
            pass
    
        try:
>           names, _, varkw, defaults, _, _, _ = inspect.getfullargspec(func)
E           AttributeError: 'module' object has no attribute 'getfullargspec'

C:\Users\nklap\PycharmProjects\mutmut\venv27\lib\site-packages\tri_declarative\__init__.py:294: AttributeError
```

@boxed any idea's what could be the issue?

**Comment by nklapste** [comment]
Locking the `tri.declarative` requirement to `3.0.0` seems to resolve the issues revolving with the python2.7 and pypy builds.

**Comment by boxed** [comment]
Yea we renamed the module due to various issues and dropped python 2 support. It's time!

**Comment by nklapste** [comment]
I've removed `tri.declarative` as a requirement in this branch. We should be good to go for some final reviews and landing this merge.

File: mutmut/__init__.py (hunk 2 above)
Comment (nklapste) [reply mutmut/__init__.py]: Restored this early return: it previously was haphazardly removed! 😬

**Comment by boxed** [comment]
Looks good to me! Merge when ready!

**Pull request merged**